cpp/tests/copying/scatter_to_tables_tests.cu

```diff
@@ -0,0 +1,330 @@
+      [](gdf_index_type row) { return static_cast<TypeParam>(row); },
+      [](gdf_index_type row) { return false; }});
+
+  std::vector<gdf_column*> v_cols(table_n_cols);
+  for(size_t i=0; i<v_colwrap.size(); i++)
+    v_cols[i]=v_colwrap[i].get();
+  
+  //input datatypes
+  cudf::table input_table{v_cols};
+ 
+  //1. No-Null
+  cudf::test::column_wrapper<gdf_index_type> scatter_map1(scatter_size, 
+  [](gdf_index_type row) {return row%2;}, 
+  [](gdf_index_type row) { return true;}); 
+ EXPECT_NO_THROW(cudf::scatter_to_tables(input_table, scatter_map1));
+  //2. All-Null
+  cudf::test::column_wrapper<gdf_index_type> scatter_map2(scatter_size, 
+  [](gdf_index_type row) {return row%2;}, 
+  [](gdf_index_type row) { return false;}); 
+ CUDF_EXPECT_THROW_MESSAGE(cudf::scatter_to_tables(input_table, scatter_map2),
+                           "Scatter map cannot contain null elements.");
+  //3. Some-Null
+  cudf::test::column_wrapper<gdf_index_type> scatter_map3(scatter_size, 
+  [](gdf_index_type row) {return row%2;}, 
+  [](gdf_index_type row) { return rand()%3==0;}); 
+ CUDF_EXPECT_THROW_MESSAGE(cudf::scatter_to_tables(input_table, scatter_map3),
+                           "Scatter map cannot contain null elements.");
+ }
+
+// Test Failure if scatter_map and input_table number of rows mismatch are not equal
+TYPED_TEST(ScatterToTablesTest, SizeMismatchTest) {
+  constexpr gdf_size_type table_n_cols{2};
+  constexpr gdf_size_type table_n_rows{1000};
+
+  //data
+  std::vector<cudf::test::column_wrapper<TypeParam>> v_colwrap(
+    table_n_cols, {table_n_rows,
+      [](gdf_index_type row) { return static_cast<TypeParam>(row); },
+      [](gdf_index_type row) { return false; }});
+
+  std::vector<gdf_column*> v_cols(table_n_cols);
+  for(size_t i=0; i<v_colwrap.size(); i++)
+    v_cols[i]=v_colwrap[i].get();
+  
+  //input datatypes
+  cudf::table input_table{v_cols};
+ 
+  //1. Equal
+  cudf::test::column_wrapper<gdf_index_type> scatter_map1(table_n_rows, 
+      [](gdf_index_type row) { return row; }, false);
+ EXPECT_NO_THROW(cudf::scatter_to_tables(input_table, scatter_map1));
+  //2. Lesser
+  cudf::test::column_wrapper<gdf_index_type> scatter_map2(table_n_rows/2, 
+      [](gdf_index_type row) { return row; }, false);
+  CUDF_EXPECT_THROW_MESSAGE(cudf::scatter_to_tables(input_table, scatter_map2),
+    "scatter_map length is not equal to number of rows in input table.");
+  //3. Greater
+  cudf::test::column_wrapper<gdf_index_type> scatter_map3(table_n_rows * 2,
+      [](gdf_index_type row) { return row; }, false);
+  CUDF_EXPECT_THROW_MESSAGE(cudf::scatter_to_tables(input_table, scatter_map3),
+    "scatter_map length is not equal to number of rows in input table.");
+ }
+
+// Test expected output if some of inputs sizes are zero.
+/*
+* c   = number of input table columns
+* r   = number of input table rows
+* sm  = number of rows in scatter_map column
+* o/p = number of output tables
+*
+* c r sm  RESULT
+* 0 * *   seg fault
```

Comment:
Since there is no bound checking in `cudf::table` for _constructor_, _num_rows_, _has_nulls_, these `cudf::table` calls causes seg fault. That's why this API input also causes seg fault.

@jrhemstad reviewed as bound checking is not required in `cudf::table` in production. 
(Note: bound checking will happen in CPU).

Reply:
Blargh.

Reply:
We should never segfault if we can avoid it. That's a bug in cudf::table, isn't it? Should open an issue.

Reply:
@karthikeyann can you give more detail as why this situation will cause a segfault?

Reply:
```
auto table = cudf::table(std::vector<gdf_columns*>(0));
table.get_column(0);
```
which calls `return _columns[index];`
Behavior for Out of bound index for operator[] of std::vector is undefined. In this case, I am seeing segmentation fault.

Reply:
> ```
> auto table = cudf::table(std::vector<gdf_columns*>(0));
> table.get_column(0);
> ```
> 
> which calls `return _columns[index];`
> Behavior for Out of bound index for operator[] of std::vector is undefined. In this case, I am seeing segmentation fault.

This is no different from doing:
```
std::vector<int> v;
v[0] = 42;
```

It just shouldn't be done.

```
auto table = cudf::table(std::vector<gdf_columns*>(0));
```

It doesn't really make sense to have a `table` with zero columns. You can have a table of _empty_ columns, but zero columns seems like it shouldn't be allowed.

Reply:
Are you saying the table constructor should disallow it @jrhemstad ?

Reply:
I believe so, but maybe there's a use-case I'm missing.

Reply:
Yes, that is good.